Bazel + BazelToolchain + BazelDeps (#3457)
conan-io/docs: reference/conanfile/tools/google/bazeldeps.rst

.. code-block:: python

    def build_requirements(self):
        self.tool_requires("my_tool/0.0.1")

    def layout(self):
        # DEPRECATED: Default generators folder will be "conan" in Conan 2.x
        self.folders.generators = "conan"
        bazel_layout(self)

    def generate(self):
        bz = BazelDeps(self)
        # generate the build-mytool/BUILD.bazel file for the tool require
        bz.build_context_activated = ["my_tool"]
        bz.generate()

Running the :command:`conan install` command, the structure created is as follows:

.. code-block:: bash

    $ conan install . -pr:b default
    .
    ├── conan
    │   ├── BUILD.bazel
    │   ├── build-my_tool
    │   │   └── BUILD.bazel
    │   └── dependencies.bzl
    └── conanfile.py

Notice that *my_tool* Bazel folder is prefixed with ``build-`` which indicates that it's being used in the build context.

Properties
++++++++++

Available since: `1.62.0 <https://github.com/conan-io/conan/releases/tag/1.62.0>`_

The following properties affect the ``BazelDeps`` generator:

- **bazel_target_name** property will define the name of the target declared in the *<REPOSITORY>/BUILD.bazel*

> - **bazel_repository_name** property will define the name of the folder where

Reviewer: unfinished sentence?

Author: Oops, yes 😭

These properties can be defined at global ``cpp_info`` level or at component level (``bazel_repository_name`` is only

> available for the global level).

Reviewer: Maybe move this to the description of the properties?

Author: It's ok for me too

that, it also generates a *dependencies.bzl* file which contains a Bazel function to load all your Conan dependencies.

The ``BazelDeps`` generator can be used by name in conanfiles:

.. code-block:: python
    :caption: conanfile.py

    class Pkg(ConanFile):
        generators = "BazelDeps"

.. code-block:: text
    :caption: conanfile.txt

    [generators]
    BazelDeps

And it can also be fully instantiated in the conanfile ``generate()`` method:

.. code:: python

    from conan import ConanFile
    from conan.tools.google import BazelDeps

    class App(ConanFile):
        settings = "os", "arch", "compiler", "build_type"
        requires = "zlib/1.2.11"

        def generate(self):
            bz = BazelDeps(self)
            bz.generate()

When the ``BazelDeps`` generator is used, every invocation of ``conan install`` will
generate several bazel files. For the *conanfile.py*
above, for example:

.. code-block:: bash

    $ conan install .
    .
    ├── BUILD.bazel
    ├── conanfile.py
    ├── dependencies.bzl
    └── zlib
        └── BUILD.bazel

Every :command:`conan install` generates these files:

> * *BUILD.bazel*: empty file which is aimed only to be together the *dependencies.bzl* one.

Reviewer: ```suggestion
* *BUILD.bazel*: An empty file aimed to be alongside the 
*dependencies.bzl* one.
```

  More information `here <https://bazel.build/concepts/build-files>`__.
* *dependencies.bzl*: this file tells your Bazel *WORKSPACE* how to load the dependencies.
* *zlib/BUILD.bazel*: contains all the targets that you can load from any of your *BUILD* files. More information in
  :ref:`conan_tools_google_bazeldeps_targets`.

Let's check the content of the files created:

.. code-block:: python
    :caption: BUILD.bazel

    # This is an empty BUILD file to be able to load the dependencies.bzl one.

> 

Reviewer: I think we can skip this as we already said that it was empty

.. code-block:: python
    :caption: dependencies.bzl

    # This Bazel module should be loaded by your WORKSPACE file.
    # Add these lines to your WORKSPACE one (assuming that you're using the "bazel_layout"):
    # load("@//conan:dependencies.bzl", "load_conan_dependencies")
    # load_conan_dependencies()

    def load_conan_dependencies():
        native.new_local_repository(
            name="zlib",
            path="/path/to/conan/package/folder/",
            build_file="/your/current/working/directory/zlib/BUILD.bazel",
        )

Given the example above, and imagining that your WORKSPACE is at the same directory, you would have to add these lines in there:

> 

Reviewer: This reads more like a tutorial than a reference for the tool... Maybe we should just enumerate the files that this generator creates, giving a general explanation of what each one of those does and then adding a section (maybe that could be just for the 2.0 docs) explaining how to use this generator or how you have to modify your project to use this generator.

Author: Yes, I know, but I thought that Conan 1.x is not going to have that example so better to have more info like that. I agree with you, and there will be an example for Conan 2.x so this is covered, but the problem was for Conan 1.x

Reviewer 2: Yes, probably leave it as-is for Conan 1.X, and improve it moving to the examples in Conan 2.0

Since Conan 1.62.0, you can put all the files generated by ``BazelDeps`` into another folder using the ``bazel_layout``:

.. code-block:: python
    :caption: conanfile.py

    from conan import ConanFile
    from conan.tools.google import BazelDeps, bazel_layout

    class App(ConanFile):
        settings = "os", "arch", "compiler", "build_type"
        requires = "zlib/1.2.11"

        def layout(self):
            # DEPRECATED: Default generators folder will be "conan" in Conan 2.x
            self.folders.generators = "conan"
            bazel_layout(self)

        def generate(self):
            bz = BazelDeps(self)
            bz.generate()

Running again the :command:`conan install` command, we now get this structure:

.. code-block:: bash

    $ conan install .
    .
    ├── conan
    │   ├── BUILD.bazel
    │   ├── dependencies.bzl
    │   └── zlib
    │       └── BUILD.bazel
    └── conanfile.py

Now your Conan-bazel files were generated in the *conan/* folder, so your WORKSPACE will look like:

.. code-block:: python
    :caption: WORKSPACE

    load("@//conan:dependencies.bzl", "load_conan_dependencies")
    load_conan_dependencies()

> Build requirements
> ++++++++++++++++++

Reviewer: Being this a reference, I think that the title should have the attribute name, that is the entity you are documenting here
```suggestion
Build build_context_activated
+++++++++++++++++++++++++++++
```
